source/common/upstream/health_discovery_service.h

```diff
@@ -133,6 +142,24 @@ class HdsDelegate : Grpc::AsyncStreamCallbacks<envoy::service::health::v3::Healt
   void onRemoteClose(Grpc::Status::GrpcStatus status, const std::string& message) override;
   envoy::service::health::v3::HealthCheckRequestOrEndpointHealthResponse sendResponse();
 
+  // Config::VersionedClient
+  const std::string methodNameTemplate() const override {
+    return STREAM_HEALTH_CHECK_METHOD_NAME_TEMPLATE;
+  }
+  const std::string
+  serviceNamespace(envoy::config::core::v3::ApiVersion transport_api_version) const override {
+    switch (transport_api_version) {
+    case envoy::config::core::v3::ApiVersion::AUTO:
+      FALLTHRU;
+    case envoy::config::core::v3::ApiVersion::V2:
+      return "discovery";
```

Comment:
So, the whole service namespace is a workaround for the fact that HDS migrated? Would it be cleaner to just package up v2/v3 method descriptors in a struct and pass them around vs. the templating?

Reply:
I'm not sure I get it. Do you think to have something like the following (e.g. for ext_authz)?

```cpp
GrpcClientImpl::GrpcClientImpl(Grpc::RawAsyncClientPtr&& async_client,
                               const absl::optional<std::chrono::milliseconds>& timeout,
                               envoy::config::core::v3::ApiVersion transport_api_version,
                               bool use_alpha)
    : async_client_(std::move(async_client)), timeout_(timeout),
      versioned_methods_("envoy.service.auth.v3.Authorization.Check",
                         "envoy.service.auth.v2.Authorization.Check",
                         "envoy.service.auth.v2alpha.Authorization.Check"),
      transport_api_version_(transport_api_version),
      service_method_(
          versioned_methods_.getMethodDescriptorForVersion(transport_api_version, use_alpha)) {}

```

I wish there is a way to list down all available services from a descriptor pool.

Reply:
Yep, it's a bit of a pain, but I think it's less confusing to make it explicit. Thanks for updating and this awesome contribution!